packages/core/src/fhir-to-cda/cda-templates/components/medications.ts

```diff
@@ -151,77 +144,115 @@ function createAugmentedMedicationStatements(
     if (!medication) {
       throw new BadRequestError("MedicationStatement must reference an existing Medication");
     }
-    return new AugmentedMedicationStatement(statement, medicationsSectionName, medication);
+    return new AugmentedMedicationStatement(medicationsSectionName, statement, medication);
   });
 }
 
 function getDosageFromMedicationStatement(statement: MedicationStatement): string {
-  const dosageValue = statement.dosage?.[0]?.doseAndRate?.[0]?.doseQuantity?.value;
-  const dosageUnit = statement.dosage?.[0]?.doseAndRate?.[0]?.doseQuantity?.unit;
-  return `${dosageValue} ${dosageUnit}`;
+  return (
+    statement.dosage
+      ?.map(dosage => {
+        return dosage.doseAndRate
+          ?.map(doseAndRate => {
+            if (doseAndRate.doseQuantity?.value && doseAndRate.doseQuantity?.unit) {
+              return `${doseAndRate.doseQuantity.value} ${doseAndRate.doseQuantity.unit}`;
+            }
+            return undefined;
+          })
+          .filter(Boolean);
+      })
```

Comment:
Cool that we're mapping all entries! I'd just recommend that we create small functions for each, like:
- a function that converts a single doseAndRate
- a function that converts a single dosage

It makes it easier to read, easier to test, and we can reuse if needed.

Reply:
Fixed